include/host/wasi_crypto/utils/optional.h

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2019-2022 Second State INC
+
+//===-- wasi_crypto/utils/handles_manager.h - OptionalRef definition ------===//
+//
+// Part of the WasmEdge Project.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains definition of OptionalRef and some helper function
+///
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "host/wasi_crypto/utils/error.h"
+
+#include <functional>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace WasmEdge {
+namespace Host {
+namespace WasiCrypto {
+
+template <typename T> using OptionalRef = T *;
+
+namespace detail {
+
+template <typename> struct IsOptional : std::false_type {};
+template <typename T> struct IsOptional<std::optional<T>> : std::true_type {
+  using Type = T;
+};
+
+template <typename> struct IsExpected : std::false_type {};
+template <typename T> struct IsExpected<WasiCryptoExpect<T>> : std::true_type {
+  using Type = T;
+};
+
+template <typename> struct IsOptionalRef : std::false_type {};
+template <typename T> struct IsOptionalRef<OptionalRef<T>> : std::true_type {
+  using Type = T;
+};
+
+template <typename> struct IsExpectedOptionalRef : std::false_type {};
+template <typename T>
+struct IsExpectedOptionalRef<WasiCryptoExpect<OptionalRef<T>>>
+    : std::true_type {
+  using Type = T;
+};
+
+} // namespace detail
+
+/// std::optional<T> -> (T -> WasiCrypto<R>) ->
+/// WasiCryptoExpect<std::optional<R>>
+template <typename O, typename F>
```

Comment:
The `enable_if_t` can be moved here to simplify the line 59-68. If you just want the `Optional` or `Function` declarations, you can use `std::declval<O>()` or `std::declval<F>()` instead. And the `forward` can also be removed in this case.